net/dispatcher.go

```diff
@@ -19,44 +23,158 @@ import (
+			return nil
+		}
+
+		if l.UseForwardProxy() {
+			proxyUrl, isValid, err := d.setProxy(httpProxy)
+			if err != nil {
+				return err
+			}
 
-	tr := &http.Transport{
-		MaxIdleConns:          100,
-		IdleConnTimeout:       10 * time.Second,
-		MaxIdleConnsPerHost:   10,
-		TLSHandshakeTimeout:   3 * time.Second,
-		ExpectContinueTimeout: 1 * time.Second,
+			if isValid {
+				d.transport.Proxy = http.ProxyURL(proxyUrl)
+			}
+		}
+
+		return nil
 	}
+}
 
-	if licenser.UseForwardProxy() {
-		proxyUrl, isValid, err := d.setProxy(httpProxy)
-		if err != nil {
+// AllowListOption sets a list of IP prefixes which will outgoing traffic will be granted access
+func AllowListOption(allowList []string) DispatcherOption {
+	return func(d *Dispatcher) error {
+		if len(allowList) == 0 {
+			return ErrAllowListIsRequired
+		}
+
+		netAllow := make([]netip.Prefix, len(allowList))
+		for i, prefix := range allowList {
+			parsed, err := netip.ParsePrefix(prefix)
+			if err != nil {
+				return fmt.Errorf("%w: %v in allowlist", ErrInvalidIPPrefix, err)
+			}
+			netAllow[i] = parsed
+		}
+
+		d.rules.Allow = netAllow
+		return nil
+	}
+}
+
+// BlockListOption sets a list of IP prefixes which will outgoing traffic will be denied access
+func BlockListOption(blockList []string) DispatcherOption {
+	return func(d *Dispatcher) error {
+		if len(blockList) == 0 {
+			return ErrBlockListIsRequired
+		}
+
+		netBlock := make([]netip.Prefix, len(blockList))
+		for i, prefix := range blockList {
+			parsed, err := netip.ParsePrefix(prefix)
+			if err != nil {
+				return fmt.Errorf("%w: %v in blocklist", ErrInvalidIPPrefix, err)
+			}
+			netBlock[i] = parsed
+		}
+
+		d.rules.Block = netBlock
+		return nil
+	}
+}
+
+// EnforceSecureOption allow self-signed certificates if set to false
+// which is susceptible to MITM attacks.
+func EnforceSecureOption(enforceSecure bool) DispatcherOption {
+	return func(d *Dispatcher) error {
+		if !enforceSecure {
+			d.transport.TLSClientConfig = &tls.Config{
+				InsecureSkipVerify: true,
+			}
+		} else {
+			d.transport.TLSClientConfig = &tls.Config{
+				MinVersion: tls.VersionTLS12,
+			}
+		}
+
+		return nil
+	}
+}
+
+func LoggerOption(logger *log.Logger) DispatcherOption {
+	return func(d *Dispatcher) error {
+		if logger == nil {
+			return ErrLoggerIsRequired
+		}
+
+		d.logger = logger
+		return nil
+	}
+}
+
+func NewDispatcher(options ...DispatcherOption) (*Dispatcher, error) {
+	d := &Dispatcher{
+		client: &http.Client{},
+		rules:  &netjail.Rules{},
+		transport: &http.Transport{
+			MaxIdleConns:          100,
+			IdleConnTimeout:       10 * time.Second,
+			MaxIdleConnsPerHost:   10,
+			TLSHandshakeTimeout:   3 * time.Second,
+			ExpectContinueTimeout: 1 * time.Second,
+		},
+	}
+
+	for _, option := range options {
+		if err := option(d); err != nil {
 			return nil, err
 		}
+	}
 
-		if isValid {
-			tr.Proxy = http.ProxyURL(proxyUrl)
+	if len(d.rules.Allow) == 0 && len(d.rules.Block) == 0 {
```

Comment:
@jirevwe What happens if only one is set? From what I'm seeing, it will eventually fail, which is not what we want. I don't think either of them should be required. It shouldn't work differently from how it works in production today. What do you think?